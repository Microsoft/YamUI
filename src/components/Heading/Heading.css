/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import 'src/css/variables/fonts.css';

.y-heading {
  display: inline;
  font-size: inherit;
  line-height: inherit;
}
.y-heading__size-1,
.y-heading__size-2,
.y-heading__size-3,
.y-heading__size-5,
.y-heading__size-6 {
  font-weight: $fontWeight__bold;
}
.y-heading__size-4 {
  font-weight: $fontWeight__normal;
}

.y-heading__size-none {
  margin-bottom: 0;
  font-weight: inherit;
}
