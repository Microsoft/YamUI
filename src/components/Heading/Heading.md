### Examples

Basic examples:

```js { "props": { "data-example": "basic" } }
<div>
  <Heading level="1">This text is rendered in an h1 tag</Heading>
  <Heading level="2">This text is rendered in an h2 tag</Heading>
  <Heading level="3">This text is rendered in an h3 tag</Heading>
  <Heading level="4">This text is rendered in an h4 tag</Heading>
  <Heading level="5">This text is rendered in an h5 tag</Heading>
  <Heading level="6">This text is rendered in an h6 tag</Heading>
</div>
```

Size overrides:

```js { "props": { "data-example": "size overrides" } }
<div>
  <Heading level="2" size="1">This h2 is rendered at the size of an h1 tag</Heading>
  <Heading level="2" size="2">This h2 is rendered at the size of an h2 tag</Heading>
  <Heading level="2" size="3">This h2 is rendered at the size of an h3 tag</Heading>
  <Heading level="2" size="4">This h2 is rendered at the size of an h4 tag</Heading>
  <Heading level="2" size="5">This h2 is rendered at the size of an h5 tag</Heading>
  <Heading level="2" size="6">This h2 is rendered at the size of an h6 tag</Heading>
  <Heading level="2" size="none">This h2 is rendered as plain text</Heading>
</div>
```
