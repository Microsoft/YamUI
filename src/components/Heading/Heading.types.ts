/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';

export type HeadingLevel = '1' | '2' | '3' | '4' | '5' | '6';
export type HeadingSize = HeadingLevel | 'none';

export interface HeadingProps extends NestableBaseComponentProps {
  /**
   * The heading level to render, i.e. h1-h6
   */
  level: HeadingLevel;

  /** Allows a heading tag at a given level to render visually as a different level,
   * or as plain inline text.
   */
  size?: HeadingSize;
}
