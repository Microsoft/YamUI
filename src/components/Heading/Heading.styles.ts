/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { fontWeights } from '../../util/styles/fonts';
import { mergeStyleSets } from '@uifabric/styling';
import { memoizeFunction } from '@uifabric/utilities';
import { HeadingSize } from './Heading';

export interface HeadingStyleProps {
  size?: HeadingSize;
}

const getFontWeight = (size?: HeadingSize) => {
  if (size === '4') {
    return fontWeights.normal;
  }

  if (size === 'none') {
    return 'inherit';
  }

  return fontWeights.bold;
};

export const getClassNames = memoizeFunction((styleProps: HeadingStyleProps) => {
  const { size } = styleProps;

  return mergeStyleSets({
    root: {
      display: 'inline',
      fontSize: 'inherit',
      lineHeight: 'inherit',
      fontWeight: getFontWeight(size),
      marginBottom: size === 'none' ? 0 : undefined,
    },
  });
});
