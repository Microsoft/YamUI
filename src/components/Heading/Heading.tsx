/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { HeadingLevel, HeadingSize } from './enums';
import './Heading.css';


const textSizes = {
  1: 'y-textSize-xxLarge',
  2: 'y-textSize-xLarge',
  3: 'y-textSize-large',
  4: 'y-textSize-large',
  5: 'y-textSize-medium',
  6: 'y-textSize-mediumSub',
};

export { HeadingLevel, HeadingSize };

export interface HeadingProps extends NestableBaseComponentProps {
  /**
   * The heading level to render, i.e. h1-h6
   */
  level: HeadingLevel;

  /** Allows a heading tag at a given level to render visually as a different level,
   * or as plain inline text.
   */
  size?: HeadingSize;
}

/**
 * A `Heading` component renders an h1-h6 tag depending on the given level. You can also
 * override its visual styling to match a different level with `size`. Set `size="none"`
 * to render the heading as unstyled inline text.
 */
export default class Heading extends React.PureComponent<HeadingProps, {}> {
  render() {
    const { children, level } = this.props;
    const TagName = `h${level}`;

    return <TagName className={this.getClasses()}>{children}</TagName>;
  }

  // Add corresponding utility textSize class so nested Icons will scale
  private getTextClass() {
    const { level, size } = this.props;
    if (size === 'none') {
      return '';
    }

    const textSize = size || level;
    return textSizes[textSize];
  }

  private getClasses() {
    const { className, size } = this.props;
    const textClass = this.getTextClass();

    const classes = ['y-heading'];
    if (size) {
      classes.push(`y-heading__size-${size}`);
    }
    if (className) {
      classes.push(className);
    }
    if (textClass) {
      classes.push(textClass);
    }

    return classes.join(' ');
  }
}
