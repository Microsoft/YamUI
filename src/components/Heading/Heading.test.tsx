/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Heading, { HeadingProps } from '.';


describe('<Heading />', () => {
  let component: ShallowWrapper<HeadingProps, {}>;

  describe('with a level', () => {
    beforeEach(() => {
      component = shallow(<Heading level="1">test content</Heading>);
    });

    it('renders its given content', () => {
      expect(component.render().text()).toEqual('test content');
    });

    it('renders the correct tag', () => {
      expect(component.find('h1').length).toBe(1);
    });

    it('contains its base className', () => {
      expect(component.hasClass('y-heading')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with a size', () => {
    beforeEach(() => {
      component = shallow(<Heading level="2" size="3">test content</Heading>);
    });

    it('renders the correct tag level', () => {
      expect(component.find('h2').length).toBe(1);
    });

    it('renders the correct size className', () => {
      expect(component.hasClass('y-heading__size-3')).toBe(true);
    });

    it('has its base className', () => {
      expect(component.hasClass('y-heading')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with size "none"', () => {
    beforeEach(() => {
      component = shallow(<Heading level="5" size="none">test content</Heading>);
    });

    it('renders the correct tag level', () => {
      expect(component.find('h5').length).toBe(1);
    });

    it('renders the size "none" className', () => {
      expect(component.hasClass('y-heading__size-none')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<Heading level="4" className="TEST_CLASSNAME">test content</Heading>);
    });

    it('includes that className', () => {
      expect(component.hasClass('TEST_CLASSNAME')).toBe(true);
    });

    it('still includes its base className', () => {
      expect(component.hasClass('y-heading')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

});
