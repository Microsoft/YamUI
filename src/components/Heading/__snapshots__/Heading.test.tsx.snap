// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> with additional className matches its snapshot 1`] = `
<h4
  className="y-heading TEST_CLASSNAME"
>
  <Block
    bottomSpacing="medium"
    textSize="large"
  >
    test content
  </Block>
</h4>
`;

exports[`<Heading /> with level 1 matches its snapshot 1`] = `
<h1
  className="y-heading"
>
  <Block
    bottomSpacing="small"
    textSize="xxLarge"
  >
    test content
  </Block>
</h1>
`;

exports[`<Heading /> with level 2 and size 3 matches its snapshot 1`] = `
<h2
  className="y-heading y-heading__size-3"
>
  <Block
    bottomSpacing="medium"
    textSize="large"
  >
    test content
  </Block>
</h2>
`;

exports[`<Heading /> with size "none" matches its snapshot 1`] = `
<h5
  className="y-heading y-heading__size-none"
>
  test content
</h5>
`;
